service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/delete/DeleteAzureNetworkStep.java

```diff
@@ -0,0 +1,82 @@
+import bio.terra.workspace.db.ResourceDao;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.resource.controlled.flight.create.CreateAzureNetworkStep;
+import bio.terra.workspace.service.workspace.model.AzureCloudContext;
+import com.azure.resourcemanager.compute.ComputeManager;
+import java.util.UUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A step for deleting a controlled Azure Network resource. This step uses the following process to
+ * actually delete the Azure Network.
+ */
+public class DeleteAzureNetworkStep implements Step {
+  private static final Logger logger = LoggerFactory.getLogger(CreateAzureNetworkStep.class);
+  private final AzureConfiguration azureConfig;
+  private final ResourceDao resourceDao;
+  private final CrlService crlService;
+  private final AzureCloudContext azureCloudContext;
+
+  private final UUID workspaceId;
+  private final UUID resourceId;
+
+  public DeleteAzureNetworkStep(
+      AzureConfiguration azureConfig,
+      AzureCloudContext azureCloudContext,
+      CrlService crlService,
+      ResourceDao resourceDao,
+      UUID workspaceId,
+      UUID resourceId) {
+    this.crlService = crlService;
+    this.resourceDao = resourceDao;
+    this.azureCloudContext = azureCloudContext;
+    this.azureConfig = azureConfig;
+    this.workspaceId = workspaceId;
+    this.resourceId = resourceId;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException {
+    var wsmResource = resourceDao.getResource(workspaceId, resourceId);
+    var network = wsmResource.castToControlledResource().castToAzureNetworkResource();
+
+    ComputeManager computeManager = crlService.getComputeManager(azureCloudContext, azureConfig);
+    var azureResourceId =
+        String.format(
+            "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Compute/network/%s", // TODO:
+            // need
+            // correct
+            // url
+            // here
+            azureCloudContext.getAzureSubscriptionId(),
+            azureCloudContext.getAzureResourceGroupId(),
+            network.getNetworkName());
+    try {
+      logger.info("Attempting to delete network " + azureResourceId);
+
+      computeManager.networkManager().networks().deleteById(azureResourceId);
+      return StepResult.getStepResultSuccess();
+    } catch (Exception ex) {
+      logger.info("Attempt to delete Azure network failed on this try: " + azureResourceId, ex);
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, ex);
+    }
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    logger.error(
+        "Cannot undo delete of Azure network resource {} in workspace {}.",
+        resourceId,
+        workspaceId);
+    // Surface whatever error caused Stairway to begin undoing.
+    return flightContext.getResult();
```

Comment:
This is OK, but not necessary. Stairway remembers the original error that triggered the UNDO. That is always the resulting error of the flight after the undo code is executed.